### PR TITLE
[Lang] Add inf and nan

### PR DESCRIPTION
**Pull request by neozhaoliang**
Related issue = #5261

### CODE REVIEW

File: python/taichi/math/mathimpl.py [hunk 1]
```diff
@@ -691,12 +691,96 @@ def inverse(mat):  # pylint: disable=R1710
...
+       >>> @ti.kernel
+       >>> def test():
+       >>>     x = vec4(nan, -nan, inf, 1)
+       >>>     isnan(x)
+       >>>
+       >>> test()
+       [1, 1, 0, 0]
+
+    Returns:
+        For each element i of the result, returns `True` if x[i] is posititve or negative floating point NaN (Not a Number) and `False` otherwise.
```
Comment (ailzhang) [inline python/taichi/math/mathimpl.py:745]: To be precise it returns 0/1 instead of True/False right?

File: tests/python/test_inf_nan.py [hunk 2]
```diff
@@ -0,0 +1,28 @@
+import taichi as ti
+from taichi.math import inf, isinf, isnan, nan
+from tests import test_utils
+
+
+def _test_inf_nan(dt):
+    ti.init(default_fp=dt)
+
+    @ti.kernel
+    def make_tests():
```
Comment (ailzhang) [inline tests/python/test_inf_nan.py:10]: nit: add a test with vector/matrix

File: python/taichi/math/mathimpl.py [hunk 3]
```diff
@@ -691,12 +691,96 @@ def inverse(mat):  # pylint: disable=R1710
...
+        y = ti.bit_cast(fx, ti.u64)
+        return (ti.cast(y >> 32, ti.u32)
+                & 0x7fffffff) + (ti.cast(y, ti.u32) != 0) > 0x7ff00000
+
+    y = ti.bit_cast(fx, ti.u32)
+    return (y & 0x7fffffff) > 0x7f800000
+
+
+@ti.func
+def vdir(ang):
```
[the remark this answers is not in the document]
Comment (neozhaoliang) [reply python/taichi/math/mathimpl.py:759]: Yes it's unrelated to inf and nan, but it's a pretty short and straight-forward function, I think it's ok to be without a test (just too lazy to open another PR :)

File: tests/python/test_inf_nan.py [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+        x = -1.0
+        assert isnan(ti.sqrt(x)) == True
+        assert isnan(inf) == isnan(1.0) == isnan(-1) == False
+        assert isinf(inf) == isinf(-inf) == True
+        assert isinf(nan) == isinf(1.0) == isinf(-1) == False
+
+    make_tests()
+
+
+@test_utils.test(default_fp=ti.f32)
```
[the remark this answers is not in the document]
Comment (neozhaoliang) [reply tests/python/test_inf_nan.py]: fixed.

File: tests/python/test_inf_nan.py (hunk 2 above)
Comment (neozhaoliang) [reply tests/python/test_inf_nan.py:10]: test added.

**Comment by taichi-gardener** [comment]
Rebase failed, please resolve the conflicts manually.

File: python/taichi/math/mathimpl.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (neozhaoliang) [reply python/taichi/math/mathimpl.py:759]: test added

File: tests/python/test_math_module.py [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+
+@pytest.mark.parametrize('dt', [ti.f32, ti.f64])
+@test_utils.test()
+def test_inf_nan_f32(dt):
+    _test_inf_nan(dt)
+
+
+@test_utils.test()
+def test_vdir():
+    ti.init()
```
Comment (ailzhang) [inline tests/python/test_math_module.py:35]: ooc: do you need to explicit call `ti.init()` here?

**Review by ailzhang — APPROVED** [review]
LGTM!

**Pull request merged**